lib/multiplexer/multiplexer_test.go

```diff
@@ -0,0 +1,285 @@
+	ports, err := utils.GetFreeTCPPorts(1)
+	c.Assert(err, check.IsNil)
+
+	listener, err := net.Listen("tcp", net.JoinHostPort("127.0.0.1", ports[0]))
+	c.Assert(err, check.IsNil)
+
+	mux, err := New(Config{
+		Listener: listener,
+	})
+	c.Assert(err, check.IsNil)
+	go mux.Serve()
+	defer mux.Close()
+
+	backend1 := &httptest.Server{
+		Listener: mux.TLS(),
+		Config: &http.Server{Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			fmt.Fprintf(w, r.RemoteAddr)
+		}),
+		},
+	}
+	backend1.StartTLS()
+	defer backend1.Close()
+
+	remoteAddr := net.TCPAddr{IP: net.ParseIP("127.0.0.1"), Port: 8000}
+	proxyLine := ProxyLine{
+		Protocol:    TCP4,
+		Source:      remoteAddr,
+		Destination: net.TCPAddr{IP: net.ParseIP("127.0.0.1"), Port: 9000},
+	}
+
+	parsedURL, err := url.Parse(backend1.URL)
+	c.Assert(err, check.IsNil)
+
+	conn, err := net.Dial("tcp", parsedURL.Host)
+	c.Assert(err, check.IsNil)
+	defer conn.Close()
+	// send proxy line first before establishing TLS connection
+	_, err = fmt.Fprintf(conn, proxyLine.String())
+	c.Assert(err, check.IsNil)
+
+	// upgrade connection to TLS
+	tlsConn := tls.Client(conn, clientConfig(backend1))
+	defer tlsConn.Close()
+
+	// make sure the TLS call succeeded and we got remote address
+	// correctly
+	out, err := utils.RoundtripWithConn(tlsConn)
+	c.Assert(err, check.IsNil)
+	c.Assert(out, check.Equals, remoteAddr.String())
+}
+
+// TestTimeout tests client timeout - client dials, but writes nothing
+// make sure server hangs up
+func (s *MuxSuite) TestTimeout(c *check.C) {
+	ports, err := utils.GetFreeTCPPorts(1)
+	c.Assert(err, check.IsNil)
+
+	listener, err := net.Listen("tcp", net.JoinHostPort("127.0.0.1", ports[0]))
+	c.Assert(err, check.IsNil)
+
+	mux, err := New(Config{
+		Listener:     listener,
+		ReadDeadline: time.Millisecond,
+	})
+	c.Assert(err, check.IsNil)
+	go mux.Serve()
+	defer mux.Close()
+
+	backend1 := &httptest.Server{
+		Listener: mux.TLS(),
+		Config: &http.Server{Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			fmt.Fprintf(w, r.RemoteAddr)
+		}),
+		},
+	}
+	backend1.StartTLS()
+	defer backend1.Close()
+
+	parsedURL, err := url.Parse(backend1.URL)
+	c.Assert(err, check.IsNil)
+
+	conn, err := net.Dial("tcp", parsedURL.Host)
+	c.Assert(err, check.IsNil)
+	defer conn.Close()
+
+	time.Sleep(5 * time.Millisecond)
+	// upgrade connection to TLS
+	tlsConn := tls.Client(conn, clientConfig(backend1))
+	defer tlsConn.Close()
+
+	// roundtrip should fail on the timeout
+	_, err = utils.RoundtripWithConn(tlsConn)
+	c.Assert(err, check.NotNil)
+}
+
+// TestUnknownProtocol make sure that multiplexer closes connection
+// with unknown protocol
+func (s *MuxSuite) TestUnknownProtocol(c *check.C) {
+	ports, err := utils.GetFreeTCPPorts(1)
+	c.Assert(err, check.IsNil)
+
+	listener, err := net.Listen("tcp", net.JoinHostPort("127.0.0.1", ports[0]))
+	c.Assert(err, check.IsNil)
+
+	mux, err := New(Config{
+		Listener: listener,
+	})
```

Comment:
from all tests, it's obvious you could use a common function to create the multiplexer:
```go
mux, err := newMux(Config{})
```
that would initialize the listener to these defaults